### PR TITLE
Pickling and Savemainsession Doc update

### LINKED ISSUE #20228: Update "Managing Python Pipeline Dependencies" page
Update https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/ with concrete examples and make it easier to use.

Additional items to cover in this page:
- pickling / managing main session
- containers to rescue how it could help with this, how to run containers locally (links?)

Imported from Jira [BEAM-9497](https://issues.apache.org/jira/browse/BEAM-9497). Original Jira may contain additional context.
Reported by: altay.

### PR DESCRIPTION
**Pull request by svetakvsundhar**
Fixes #20228 by documenting pickling and savemainsession on Beam Docs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by svetakvsundhar** [comment]
R: @Abacn

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 1]
```diff
@@ -141,3 +141,13 @@ However, it may be possible to pre-build the SDK containers and perform the depe
 Dataflow, see [Pre-building the python SDK custom container image with extra dependencies](https://cloud.google.com/dataflow/docs/guides/using-custom-containers#prebuild).
 
 **NOTE**: This feature is available only for the `Dataflow Runner v2`.
+
+## Pickling and Managing Main Session
```
Comment (AnandInguva) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:145]: Note that this is true when the pickler is `dill`.

**Comment by svetakvsundhar** [comment]
Run Website PreCommit

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 2]
```diff
@@ -141,3 +141,12 @@ However, it may be possible to pre-build the SDK containers and perform the depe
...
+
+Pickling in the Python SDK is set up to pickle the state of the global namespace. By default, global imports, functions, and variables defined in the main session are not saved during the serialization of a Dataflow job.
+Thus, one might encounter unexpected `NameErrors` when running a `DoFn` on Dataflow Runner. To resolve this, manage the main session by
```
Comment (AnandInguva) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:148]:
```suggestion
Thus, one might encounter unexpected `NameError`s when running a `DoFn` on Dataflow Runner. To resolve this, manage the main session by
```

**Review by AnandInguva — APPROVED** [review]

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 3]
```diff
@@ -141,3 +141,12 @@ However, it may be possible to pre-build the SDK containers and perform the depe
...
 **NOTE**: This feature is available only for the `Dataflow Runner v2`.
+
+## Pickling and Managing Main Session
+
+Pickling in the Python SDK is set up to pickle the state of the global namespace. By default, global imports, functions, and variables defined in the main session are not saved during the serialization of a Dataflow job.
+Thus, one might encounter an unexpected `NameError` when running a `DoFn` on Dataflow Runner. To resolve this, manage the main session by
+simply setting `--save_main_session=True`. This will load the pickled state of the global namespace onto the Dataflow workers.
+For more information, see [Handling NameErrors](https://cloud.google.com/dataflow/docs/guides/common-errors#how-do-i-handle-nameerrors).
+
+**NOTE**: This strictly applies to the `Python SDK executing with the dill pickler on the Dataflow Runner`.
```
Comment (damccorm) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:152]: Doesn't this at least apply to all remote runners using portability (not just Dataflow)?

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md [hunk 4]
```diff
@@ -141,3 +141,12 @@ However, it may be possible to pre-build the SDK containers and perform the depe
 Dataflow, see [Pre-building the python SDK custom container image with extra dependencies](https://cloud.google.com/dataflow/docs/guides/using-custom-containers#prebuild).
 
 **NOTE**: This feature is available only for the `Dataflow Runner v2`.
+
+## Pickling and Managing Main Session
```
Comment (damccorm) [inline website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:145]: If we're going to use Dataflow specific language here, we should specifically call that out in the section heading. I think this applies to other remote runners though.

File: website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md (hunk 3 above)
Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:152]: For example, we use it in our flink portability tests - https://github.com/apache/beam/blob/326373715e0ca071d610a03e92626b1957253f81/runners/portability/test_flink_uber_jar.sh#L24

Comment (damccorm) [reply website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:152]: Also, I'd make it clear that the dill pickler is the default.

Comment (svetakvsundhar) [reply website/www/site/content/en/documentation/sdks/python-pipeline-dependencies.md:152]: ah thanks for the catch! updating to make it clearer -- this doesn't apply on `DirectRunner`.

**Review by damccorm — APPROVED** [review]
LGTM, thanks

**Pull request merged**